File: go/crypto/e2ee/framing.go
package e2ee

import (
	"errors"
	"fmt"

	"github.com/floegence/flowersec/internal/bin"
)

const (
	handshakeHeaderLen = 4 + 1 + 1 + 4
	recordHeaderLen    = 4 + 1 + 1 + 8 + 4
)

var (
	// ErrInvalidMagic indicates the frame prefix is not recognized.
	ErrInvalidMagic = errors.New("invalid magic")
	// ErrInvalidVersion signals the wire protocol version mismatched.
	ErrInvalidVersion = errors.New("invalid version")
	// ErrInvalidLength indicates a malformed or truncated frame length.
	ErrInvalidLength = errors.New("invalid length")
)

// EncodeHandshakeFrame wraps a JSON payload with the handshake header.
func EncodeHandshakeFrame(handshakeType uint8, payloadJSON []byte) []byte {
	out := make([]byte, handshakeHeaderLen+len(payloadJSON))
	copy(out[:4], []byte(HandshakeMagic))
	out[4] = ProtocolVersion
	out[5] = handshakeType
	bin.PutU32BE(out[6:10], uint32(len(payloadJSON)))
	copy(out[10:], payloadJSON)
	return out
}

// DecodeHandshakeFrame validates and extracts a handshake frame.
func DecodeHandshakeFrame(frame []byte, maxPayload int) (handshakeType uint8, payloadJSON []byte, err error) {
	if len(frame) < handshakeHeaderLen {
		return 0, nil, ErrInvalidLength
	}
	if string(frame[:4]) != HandshakeMagic {
		return 0, nil, ErrInvalidMagic
	}
	if frame[4] != ProtocolVersion {
		return 0, nil, ErrInvalidVersion
	}
	handshakeType = frame[5]
	n := int(bin.U32BE(frame[6:10]))
	if n < 0 || n > len(frame)-handshakeHeaderLen {
		return 0, nil, ErrInvalidLength
	}
	if maxPayload > 0 && n > maxPayload {
		return 0, nil, fmt.Errorf("handshake payload too large: %w", ErrInvalidLength)
	}
	if handshakeHeaderLen+n != len(frame) {
		return 0, nil, ErrInvalidLength
	}
	return handshakeType, frame[10:], nil
}

// LooksLikeRecordFrame quickly checks a frame header without decrypting.
func LooksLikeRecordFrame(frame []byte, maxCiphertext int) bool {
	if len(frame) < recordHeaderLen {
		return false
	}
	if string(frame[:4]) != RecordMagic {
		return false
	}
	if frame[4] != ProtocolVersion {
		return false
	}
	n := int(bin.U32BE(frame[14:18]))
	if n < 0 {
		return false
	}
	if maxCiphertext > 0 && n > maxCiphertext {
		return false
	}
	return recordHeaderLen+n == len(frame)
}

File: ts/src/e2ee/framing.ts
import { concatBytes, readU32be, readU64be, u32be, u64be } from "../utils/bin.js";
import { HANDSHAKE_MAGIC, PROTOCOL_VERSION, RECORD_MAGIC } from "./constants.js";

const te = new TextEncoder();

export const HANDSHAKE_HEADER_LEN = 4 + 1 + 1 + 4;
export const RECORD_HEADER_LEN = 4 + 1 + 1 + 8 + 4;

export class FramingError extends Error {}

export function encodeHandshakeFrame(handshakeType: number, payloadJsonUtf8: Uint8Array): Uint8Array {
  const header = new Uint8Array(HANDSHAKE_HEADER_LEN);
  header.set(te.encode(HANDSHAKE_MAGIC), 0);
  header[4] = PROTOCOL_VERSION;
  header[5] = handshakeType & 0xff;
  header.set(u32be(payloadJsonUtf8.length), 6);
  return concatBytes([header, payloadJsonUtf8]);
}

export function decodeHandshakeFrame(
  frame: Uint8Array,
  maxPayloadBytes: number
): { handshakeType: number; payloadJsonUtf8: Uint8Array } {
  if (frame.length < HANDSHAKE_HEADER_LEN) throw new FramingError("handshake frame too short");
  if (new TextDecoder().decode(frame.slice(0, 4)) !== HANDSHAKE_MAGIC) throw new FramingError("bad handshake magic");
  if (frame[4] !== PROTOCOL_VERSION) throw new FramingError("bad handshake version");
  const handshakeType = frame[5]!;
  const n = readU32be(frame, 6);
  if (maxPayloadBytes > 0 && n > maxPayloadBytes) throw new FramingError("handshake payload too large");
  if (HANDSHAKE_HEADER_LEN + n !== frame.length) throw new FramingError("handshake length mismatch");
  return { handshakeType, payloadJsonUtf8: frame.slice(HANDSHAKE_HEADER_LEN) };
}

export function looksLikeRecordFrame(frame: Uint8Array, maxCiphertextBytes: number): boolean {
  if (frame.length < RECORD_HEADER_LEN) return false;
  if (new TextDecoder().decode(frame.slice(0, 4)) !== RECORD_MAGIC) return false;
  if (frame[4] !== PROTOCOL_VERSION) return false;
  const n = readU32be(frame, 14);
  if (maxCiphertextBytes > 0 && n > maxCiphertextBytes) return false;
  return RECORD_HEADER_LEN + n === frame.length;
}

export function encodeU64beBigint(n: bigint): Uint8Array {
  return u64be(n);
}

export function decodeU64beBigint(buf: Uint8Array, off: number): bigint {
  return readU64be(buf, off);
}

